### ISSUE TITLE
GRPC-Web occasionally not sending responses back to clients

### ISSUE THREAD

**Issue by leosunmo** [envoyproxy/envoy#17722]
**GRPC-Web connections occasionally get in to a broken state**

First things first, this environment is running Istio 1.9.6 with Envoy version `172db4cfdf037bc8cf1613969f94c25b6198bc4f/1.17.1/Clean/RELEASE/BoringSSL`

Occasionally an Envoy sidecar seems to get in to a state where it will silently (?) drop GRPC-Web responses instead of sending the GRPC -> HTTP/2 responses back to the client/browser. The symptoms are that the browser is just continuously trying to send messages every 10 seconds and timing out. On our application side (Go GRPC server) we are getting `context cancelled` errors when we are trying to send messages on the "broken" GRPC stream *AFTER* ~10 seconds, when the browser closes the connection. Before this the GRPC Server is happily sending messages on the stream, they're just not getting to the client/browser.

The high-level architecture is this:
Browser -> Istio Ingress (Envoy) -> Sidecar (Envoy) -> Go GRPC server

We only run *ONE* replica of the Go GRPC server, so all GRPC-Web connections go to this one server.

We have *not* been able to _reliably_ reproduce this yet. This happens maybe 4-5 times in a week in our fairly low-traffic development environment, but more frequently in the production environment, but we are having issues tracking this issue, so unsure how frequent it is. Our solution to capture information in our development environment has been to turn everything to Debug and analyse everything for a week.

## The browser side
We first figured this would be a GRPC-Web browser-side issue, so started capturing information there.

We managed to capture some netlog from a Chrome browser that was in this "broken" state. The "broken" state is on the left and a normal successful connection on the right as comparison:
![image-20210811-085425 (1)](https://user-images.githubusercontent.com/13341934/129494295-6959f75e-b350-4aca-b57e-2c4ca95561a2.png)

As you can see it never receives any data after it receives the headers from Envoy/server, where as in the successful one it immediately receives data after the headers.

When a browser gets in to this state, you can to completely restart it or do some kind of hard refresh. Just refreshing the page doesn't seem to help.

## The Envoy side
The Envoy Access logs and Debug logs didn't show anything obvious, as Envoy (correctly) thinks these connections have been gracefully and correctly closed. So everything is `200` / `0` with no error flags. 

However, if we filter on any connection that lives for about 10 seconds, we can see a pattern. This also revealed that specific Connection IDs seem to have the issue. 

For example if connection ID `C150` has the issue and 2 different clients/browsers are connection through this connection ID, they will both have their responses from the Go GRPC server dropped.

![image](https://user-images.githubusercontent.com/13341934/129495055-612e7e60-da1f-4081-93a0-789ff963fac4.png)
This is not the best evidence, but on the left is the Envoy Stream Reset messages from the "broken" connection ID in one of the timespans that had the issue, and on the right is the Go GRPC server context cancelled messages, all with a duration of ~`12,000` milliseconds. The extra 2 seconds is because of the healthcheck message being sent every 4 seconds, so doesn't catch the stream closing at 10 seconds, but instead notices at 12 seconds when it tries to send on the stream.

Since we can't easily reproduce this I haven't been able to run this with Envoy at a `trace` level yet. It running in Istio doesn't help either as I'm not entire sure how to run it with `-l trace`.

Let me know what else I can provide here, or **any** potential workarounds as this is affecting us in production right now.

**Comment by leosunmo** [comment]
To add some information to this, we just observed the same issue with a NodeJS application running in an AWS Lambda.

This would further point towards Envoy being the issue here. The NodeJS application has a 30 second timeout, and the Go GRPC server deemed the connection "healthy" for 30 seconds until the NodeJS client closed it because of timeout since it wasn't receiving any data from the Go server after the initial headers.

**Comment by mattklein123** [comment]
Please test with a more recent version of Envoy. 1.17.1 is quite old and it's possible whatever issue this is has been fixed.

**Comment by leosunmo** [comment]
We upgraded to `1.19.0` a day ago and I will report back.

**Comment by mattklein123** [comment]
> We upgraded to 1.19.0 a day ago and I will report back.

Cool, thanks. We will debug further if it still repros.

**Comment by leosunmo** [comment]
> > We upgraded to 1.19.0 a day ago and I will report back.
> 
> Cool, thanks. We will debug further if it still repros.

This is occurring right now on a connection to a 1.19 Envoy. I've enabled Trace-level logs trying to collect as much as I can.

**Comment by leosunmo** [comment]
I'm collecting a lot of logs, but nothing obvious is standing out to me, and at some point we will have more noise than signal :(

**Comment by mattklein123** [comment]
Unfortunately I'm not sure. My guess is some type of bug within the grpc-web filter but it's hard to say.

cc @alyssawilk @markdroth do you know if there is anyone on the gRPC team that would be willing to help debug this?

**Comment by markdroth** [comment]
@wenbozhu, can you find someone to help out with this grpc-web issue?

**Comment by wenbozhu** [comment]
>  we just observed the same issue with a NodeJS application running in an AWS Lambda.

I assume that the NodeJS application talks to Envoy via grpc (v.s. grpc-web). If so, the bug shouldn't be in the grpc-web filter?

**Comment by leosunmo** [comment]
No, it's a browser talking through Envoy to a Go and NodeJS backend. The fact that it's happening for both a NodeJS-based backend and the Go backend (kind of) rules out the Go GRPC library having a bug. Also the fact that all of our GRPC traffic is not having this problem, only GRPC-web. We have more GRPC traffic than we have GRPC-Web traffic, so it's not a volume issue either.

You can also see in the netlog that the Browser side of things seems to be working fine. It just doesn't receive any data after the initial headers and sits there until it (presumably the GRPC-Web Javascript library) times out and sends a RST_STREAM back to Envoy/Backend.

**Comment by leosunmo** [comment]
I have a trace-level capture filtered on a specific connection ID that was exhibiting this behaviour a few days ago.
[trace-C80317.csv](https://github.com/envoyproxy/envoy/files/7034925/trace-C80317.csv)
It's a CSV because that's what Elasticsearch exported :shrug: But it's only an extra timestamp as the first field, and then the normal Envoy logs, so just open it in a text editor and ignore the first timestamp.

I couldn't really find anything of interests in here but these connections were exhibiting the "no data going back to browser" behaviour during the time of the log capture.

**Comment by leosunmo** [comment]
This is more and more starting to look like it might be connected to https://github.com/istio/istio/issues/34973. We see similar "bad sessions" being established with good old RESTful HTTP connections. A single user/browser ends up in a situation where the replies from our upstreams are just being silently dropped somewhere, and the ingress gateway retries and retries and eventually gives up and `UT, URX`, and 504's the client.

This _looks_ like if it was happening to web-grpc as well, it'd look like what I described in this ticket. So it's probably the same issue. Any ideas here? Should I redo a ticket in Envoy for this issue, modify this issue?

**Comment by alyssawilk** [comment]
I think largely we'd need more data to look into this :-/
This is a multi-system issue with no reproducability or even clarity which server is causing the problem.  If there were traces of Envoy reading in TCP data and not processing it, or reading stream data and not forwarding it, we could look into it but absent more information about what's going on there's not much we could do at the moment.

**Comment by leosunmo** [comment]
@alyssawilk appreciate you taking the time, even if there's no progress! I completely agree that there's little we can do if we can't reproduce it reliably, you can imagine my frustration... It's still happening with some frequency, but I am pretty sure it's not specifically GRPC/GRPC-Web related, but some broader issue with connections. We are trying to work around this with limited success.

Would it be helpful to try and `tcpdump` our sidecar Envoy if I catch it happening again?

**Comment by alyssawilk** [comment]
I hear you - intermittent issues are the worst to debug.
I guess if I were trying to track this down I'd try a couple of things.  I'd try to come up with a repro case, with an Envoy set up in my custom configuration and varied payloads.  If I were sure the connection was around and the data was not processed I'd focus on proxying different size payloads and maybe even tweaking network or kernel buffer conditions to try to force it to wedge and capture Envoy logs and tcpdump for debug.  If it could be a connection close issue I'd be more focused on response code details, seeing if there was a spike in an unusual error response cause when I saw the issues.
